File: June-LeetCoding-Challenge-2020/Week4/96-unique-binary-search-tree.py
class Solution:
    # Sol 1. Catalan Number
    def numTrees(self, n: int) -> int:
        if n is 0:
            return 0
        elif n is 1:
            return 1
        return self.catalan_number(n)
 
    def catalan_number(self, c:int) -> int:
        if c<=1:
            return 1
        re = 0
        re = math.factorial(2*c)//(math.factorial(c+1)*math.factorial(c))
        return re
    
    # TC: O(n)
    # SC: O(1)
    # Runtime: 32 ms, faster than 47.87% 
    # Memory Usage: 13.9 MB, less than 38.13% 

    # Sol 2. divide and conquer
    def numTrees(self, n: int) -> int:
        if n == 0:
            return 0
        elif n == 1:
            return 1  
        mem = {}  # memoization
        def d_c(start,end): # divide and conquer
            if start<1 or end>n or start>=end:
                return 1
            count = 0
            if (start, end) in mem:
                return mem[(start, end)]
            else:
                for i in range(start, end+1):
                    count += d_c(start, i-1)*d_c(i+1, end)
                mem[(start,end)] = count
            return count
        
        return d_c(1,n) 
    # TC: O(n*h), n is number of nodes, h is depth of each tree, worst case n^2
    # SC: O(n), map m is declared in the scope
    # Runtime: 28 ms, faster than 76.04%
    # Memory Usage: 13.8 MB, less than 50.56%  

File: June-LeetCoding-Challenge-2020/Week4/96-unique-binary-search-tree.js
/**
 * @param {number} n
 * @return {number}
 * FAILED SOLUTION :[
 */
var numTrees = function(n) {
    if (n <= 1) {
        return n;
    }

    return countBST(Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER, n);
};

function countBST (min, max, used, size) {
    if (used.size === size) {
        return 1;
    }

    let count = 0;

    for (let num = 1; num <= size; num++) {

        if (!used.has(num) && num > min && num < max) {

            used.add(num);
            const left = countBST(min, num, used, size);
            const right = countBST(num, max, used, size);
            used.delete(num);

            if (left !== 0 && right !== 0) {

                count += (left * right);
            } else if (left === 0) {

                count += right;
            } else if (right === 0) {

                count += left;
            }
        }
    }

    return count;
}

/**
 * @param {number} n
 * @return {number}
 * Runtime: 3312 ms, faster than 6.50% of JavaScript online submissions for Unique Binary Search Trees.
 * Memory Usage: 34.4 MB, less than 8.38% of JavaScript online submissions for Unique Binary Search Trees.
 * TC: O(2^n)
 * SC: O(n)
 */
var numTrees = function(n) {
    if (n <= 1) {
        return 1;
    }

    let count = 0;
    for (let num = 1; num <= n; num++) {
        count += numTrees(num - 1) * numTrees(n - num);
    }

    return count;
};

/**
 * @param {number} n
 * @return {number}
 * Runtime: 68 ms, faster than 44.51% of JavaScript online submissions for Unique Binary Search Trees.
 * Memory Usage: 32.7 MB, less than 96.89% of JavaScript online submissions for Unique Binary Search Trees.
 * TC: O(n * k), where k is depth of BST which veries from logn to n -> O(n^2)
 * SC: O(n)
 */
var numTrees = function(n, memo = new Map()) {
    if (n <= 1) {
        return 1;
    } else if (memo.has(n)) {
        return memo.get(n);
    }

    let count = 0;
    for (let num = 1; num <= n; num++) {
        count += numTrees(num - 1, memo) * numTrees(n - num, memo);
    }

    memo.set(n, count);

    return count;
};